db/index.go

```diff
@@ -11,120 +11,213 @@ import (
 	"fmt"
 	"sort"
 
+	"github.com/alecthomas/jsonschema"
 	ds "github.com/ipfs/go-datastore"
 	"github.com/ipfs/go-datastore/query"
 	"github.com/tidwall/gjson"
 	"github.com/tidwall/sjson"
 )
 
-// size of iterator keys stored in memory before more are fetched
 const (
+	// iteratorKeyMinCacheSize is the size of iterator keys stored in memory before more are fetched.
 	iteratorKeyMinCacheSize = 100
 )
 
-// ErrUniqueExists is the error thrown when data is being inserted for a unique constraint value that already exists
 var (
-	indexPrefix     = ds.NewKey("_index")
+	// ErrUniqueExists indicates an insert resulted in a unique constraint violation.
 	ErrUniqueExists = errors.New("unique constraint violation")
+	// ErrNotIndexable indicates an index path does not resolve to a value.
 	ErrNotIndexable = errors.New("value not indexable")
-	ErrNoIndexFound = errors.New("no index found")
-)
+	// ErrCantCreateUniqueIndex indicates a unique index can't be created because multiple instances share a value at path.
+	ErrCantCreateUniqueIndex = errors.New("can't create unique index (duplicate instances exist)")
+	// ErrIndexNotFound indicates a requested index was not found.
+	ErrIndexNotFound = errors.New("index not found")
 
-// Indexer is the interface to implement to support Collection indexes
-type Indexer interface {
-	BaseKey() ds.Key
-	Indexes() map[string]Index //[indexname]indexFunc
-}
+	indexPrefix = ds.NewKey("_index")
+	indexTypes  = []string{"string", "number", "integer", "boolean"}
+)
 
-// Index is a function that returns the indexable, encoded bytes of the passed in bytes
+// Index defines an index.
 type Index struct {
-	IndexFunc func(name string, value []byte) (ds.Key, error)
```

Comment:
Fair enough, though it would be nice to be able to index certain field types differently? I guess we could encompass that in a single function still...

Reply:
Yep yep